### DISCUSSION TITLE
Netty channel read callback and channel future listener callback order of invocation

### DISCUSSION THREAD

**Discussion by dileepmandapam** [netty/netty#14541] [Q&A] [answered]
I have a question on order of callbacks invocation. Client is sending a request to the server by invoking writeAndFlush and logging once the request sent successfully.
`
```java
		ch.writeAndFlush(convertToByteBuffer(requestInJson)).addListener((ChannelFutureListener) channelFuture -> {
                    if (channelFuture.isSuccess()) {
                        logger.info("Successfully sent request.");
                    }
 });
```
using channelRead0 callback, client is processing the response messages from the server.
```java
protected void channelRead0(ChannelHandlerContext ctx, String msg) throws Exception {
     logger.info("Received response :{}", msg);
}
``` 
The question is on the order of invocation of these callbacks. In most of the cases, the writeAndFlush callback getting invoked first and then channelRead0 callback getting invoked. 
```java
Successfully sent request.
Received response : adadasda
```
But for a few cases, the order is reverse. 
```java
Received response : adadasda

Successfully sent request.

```

First of all why the order is messed up. Can we enforce the order of callbacks ?.

**Comment by chrisvest** [comment, marked as answer]
When a promise is completed, such as when a write is flushed, then its attached ChannelFutureListeners gets sent to the event loop as tasks.

There's no guarantee between this callback, and observing the readable data on the network, because the event loop intermixes IO and task execution.

**Comment by dileepmandapam** [comment]
Thanks !!. Removed the dependency order in the callbacks
